pkg/relayer/proxy/relay_meter.go

```diff
@@ -0,0 +1,383 @@
+package proxy
+
+import (
+	"context"
+	"math/big"
+	"strings"
+	"sync"
+
+	"cosmossdk.io/depinject"
+	"cosmossdk.io/math"
+	cosmostypes "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/gogoproto/proto"
+
+	"github.com/pokt-network/poktroll/app/volatile"
+	"github.com/pokt-network/poktroll/pkg/client"
+	"github.com/pokt-network/poktroll/pkg/client/tx"
+	"github.com/pokt-network/poktroll/pkg/crypto/protocol"
+	"github.com/pokt-network/poktroll/pkg/either"
+	"github.com/pokt-network/poktroll/pkg/observable"
+	"github.com/pokt-network/poktroll/pkg/observable/channel"
+	"github.com/pokt-network/poktroll/pkg/relayer"
+	apptypes "github.com/pokt-network/poktroll/x/application/types"
+	servicetypes "github.com/pokt-network/poktroll/x/service/types"
+	sessionkeeper "github.com/pokt-network/poktroll/x/session/keeper"
+	sessiontypes "github.com/pokt-network/poktroll/x/session/types"
+	sharedtypes "github.com/pokt-network/poktroll/x/shared/types"
+)
+
+var _ relayer.RelayMeter = (*ProxyRelayMeter)(nil)
+
+// appRelayMeter is the relay meter's internal representation of an application's
+// max and consumed stake.
+type appRelayMeter struct {
+	// The onchain application the relay meter is for.
+	app apptypes.Application
+	// The maximum uPOKT an application can pay this relayer for a given session.
+	maxCoin cosmostypes.Coin
+	// The amount of uPOKT a specific application has consumed from this relayer in the given session.
+	consumedCoin cosmostypes.Coin
+	// The current sessionHeader the application is metered in.
+	sessionHeader *sessiontypes.SessionHeader
+
+	// sharedParams, service and serviceRelayDifficulty are used to calculate the relay cost
+	// that increments the consumedAmount.
+	// They are cached at each session to avoid querying the blockchain for each relay.
+	// TODO_TECHDEBT(#543): Remove once the query clients start handling caching and invalidation.
+	sharedParams           *sharedtypes.Params
+	service                *sharedtypes.Service
+	serviceRelayDifficulty servicetypes.RelayMiningDifficulty
+}
+
+// ProxyRelayMeter is the offchain Supplier's rate limiter.
+// It ensures that no Application is over-serviced by the Supplier per session.
+// This is done by maintaining the max amount of stake the supplier can consume
+// per session and the amount of stake consumed by mined relays.
+type ProxyRelayMeter struct {
+	// supplierToAppMetricsMap is a map of supplier addresses to application address
+	// to the application's relay meter.
+	// Only known applications (i.e. have already requested relaying) that have their stakes metered.
+	// This map gets reset every new session in order to meter new applications, since the old
+	// ones might have another Supplier set for their sessions.
```

Comment:
Do we need to be worried/concerned/consider grace periods?

Reply:
No worries about grace periods they are always shorter than settlement duration.